etcd cluster expansion

slaperche-scality opened the pull request:
How to test:

See procedure [here](https://docs.google.com/document/d/1GAMbSTIFhCypBUITg_SunCOeoMdbEfBK_XZZJIw3Wdw/edit).

slaperche-scality committed:
salt: populate the mine with etcd endpoints

Add a function that push endpoints if an etcd is running on the current
node.

By using this function in the mine, we have an automatically up-to-date
list of etcd endpoints, which is useful to extend the etcd cluster
(since we need to provide a value for `initial-cluster`.

Refs: #738
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/modules: add a function to check if a compopent is running

For now, we only check if the a pod is listed by `crictl pods` and has
the specified component.
This can certainly be improved…

Refs: #738
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/738-etcd-cluster-expansion` into
`development/2.0`
* [w/2.1/feature/738-etcd-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/738-etcd-cluster-expansion) will be merged into `development/2.1` 
* [w/2.2/feature/738-etcd-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/738-etcd-cluster-expansion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

sayf-eddine-scality commented on salt/_modules/cri.py:
```diff
@@ -148,3 +148,22 @@ def wait_container(name, state, timeout=60, delay=5):
+       correctly on the system, e.g. in :file:`/etc/crictl.yaml`.
+    '''
+    log.info('Checking if compopent %s is running', name)
+    out = __salt__['cmd.run_all']('crictl pods -o json')
```
wouldnt `crictl ps` be more exact as the pod may exist but it is in a crash loop

slaperche-scality replied:
Used the `crictl pods` but with extra params, as suggested by Nicolas.

jbertran commented on salt/_modules/cri.py:
```diff
@@ -148,3 +148,22 @@ def wait_container(name, state, timeout=60, delay=5):
+
+
+def component_is_running(name):
+    '''Return true the specified component is running.
```
+if

NicolasT commented on salt/_modules/cri.py:
```diff
@@ -148,3 +148,22 @@ def wait_container(name, state, timeout=60, delay=5):
+        log.error('Failed to list pods')
+        return False
+    for pod in salt.utils.json.loads(out['stdout'])['items']:
+        if pod['labels'].get('component') == name:
```
Just do the filtering on the CLI? `crictl pods --label component=etcd --state=ready -o json`

slaperche-scality replied:
Ho, neat!

NicolasT commented on salt/_modules/metalk8s.py:
```diff
@@ -0,0 +1,32 @@
+        return
+    if __salt__['cri.component_is_running'](name='etcd'):
+        ips = __salt__['network.ip_addrs'](cidr=cidr)
+        if isinstance(ips, list):
```
What else can it be?

slaperche-scality replied:
Indeed, that seems useless (I copied that straight from a draft of Teddy's branch).

NicolasT commented on salt/_modules/metalk8s.py:
```diff
@@ -0,0 +1,32 @@
+        if isinstance(ips, list):
+            return [
+                'https://{ip}:{port}'.format(ip, DEFAULT_ETCD_PORT)
+                for ip in ips
```
Nope, we bind `etcd` to only one of them.

Overall, given #757, I'd like us to have all servers to pick *one* control-plane IP and *one* workload-plane IP, then publish the chosen IPs in the mine, and bind services to only this IP, after which any client of such service(s) can retrieve its address easily, and have it uniquely identified.

Furthermore, over time, it'd allow us to recreate client configuration when relevant values in the mine change by using a reactor (e.g. when a server IP is changed).

All of the above is, by the way, not restricted to `etcd` ;-)

slaperche-scality replied:
Ok, then I'll always pick the first one for now.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

slaperche-scality committed:
salt/orch: add runners to deploy a new etcd node

For now we only generate the certificates for etcd on the new node.

Refs: #738
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# History mismatch



Merge commit #160fe53fe8c4c244200a1ff0fe745dbcc1c698ce on the integration branch
`w/2.1/feature/738-etcd-cluster-expansion` is merging a branch which is neither the current
branch `feature/738-etcd-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/738-etcd-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality committed:
salt/etcd: ensure that the etcd CA is present

It is required to start `etcd`, as value for:
- `--peer-trusted-ca-file`
- `--trusted-ca-file`

Refs: #738
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/orch: force a refresh of the mine

To be sure to have an up-to-date list of `etcd` node we force a refresh
of the mine before deploying the new `etcd` node.

Refs: #738
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/etcd: update the etcd deployment state

Now we fetch the list of etcd endpoints to decide if we create a new
cluster (no existing endpoints) or join an existing one.

Thanks to the list of endpoints we can compute the value for
`initial-cluster`.

Refs: #738
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/etcd: advertise the etcd endpoint in the mine

Once we are sure that the etcd node is up, we publish its endpoints in
the Salt mine.

Refs: #738
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/orch: deploy updated etd manifest

Refs: #738
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

NicolasT commented on salt/_modules/cri.py:
```diff
@@ -148,3 +148,21 @@ def wait_container(name, state, timeout=60, delay=5):
+    '''
+    log.info('Checking if compopent %s is running', name)
+    out = __salt__['cmd.run_all'](
+        'crictl pods --label component=etcd --state=ready -o json'
```
Shouldn't this use `name`?

slaperche-scality replied:
>_<, of course

NicolasT commented on salt/metalk8s/kubeadm/init/etcd/local.sls:
```diff
@@ -7,6 +7,24 @@
 
+{% set endpoint  = host_name ~ '=https://' ~ host ~ ':2380' %}
+
+{%- set ca_cert = salt['mine.get']('*', 'kubernetes_etcd_ca_server').values()[0] %}
```
`kubernetes_etcd_ca_server` is a bit of an odd name for what is actually the CA certificate itself (if I got this correctly). Not in scope of this PR to fix it, I guess, but nonetheless, something to reconsider.

slaperche-scality replied:
Tracked by #890

NicolasT reviewed (commented):
Mostly LGTM for now.

slaperche-scality committed:
salt/modules: add helper for `etcdctl member add`

This new function run, in the `etcd` pod, the required `etcdctl member
add` command to add a new `etcd` node into the `etcd` cluster.

Refs: #738
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/orch: register the node in the etcd cluster

Closes: #738
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

gdemonet commented on salt/metalk8s/kubeadm/init/etcd/local.sls:
```diff
@@ -8,9 +8,24 @@
+{%- set etcd_endpoints = salt.saltutil.runner('mine.get', tgt='*', fun='etcd_endpoints').values() %}
+
+{#- Compute the initial state according to the existing list of node. #}
+{%- if etcd_endpoints %}
```
Not sure if clearer, but you could use `{%- set state = "existing" if etcd_endpoints else "new" -%}`

slaperche-scality replied:
Looks clearer to me, thanks!

gdemonet commented on salt/_modules/metalk8s.py:
```diff
@@ -29,3 +44,42 @@ def get_etcd_endpoint():
+    pod_name = 'etcd-{}'.format(__salt__['network.get_hostname']())
+    try:
+        client = kubernetes.config.new_client_from_config(
+            config_file='/etc/kubernetes/admin.conf'
```
This file is only available on the bootstrap node (and Salt master). It would be interesting to mention this in the docstring - e.g. "this method should only run on the Salt master".

slaperche-scality replied:
Yup, indeed it's worth to mention it.

TeddyAndrieux replied:
If it's only to run on the `salt master` you should use a runner not a module. 
Here it's works because we use an orchestrate so we can run "modules" on the master but normally it's only runner who run on the master.

https://docs.saltstack.com/en/latest/ref/runners/

slaperche-scality replied:
So where should I put this code?

TeddyAndrieux replied:
Since it's a runner in `salt/_runners/xxx.py`

NicolasT replied:
I disagree, @TeddyAndrieux. This is an action that *could* be performed on any node (and, indeed, at some point we may decide to actually run it on a non-master minion, orchestrated by a runner). As such, it's a module. We simply happen to invoke it on the master for now.

Here's how I understand the Salt taxonomy:

- Module procedure: and 'action', something to be executed somewhere (likely a minion), very imperative.
- State procedure: code which exposes functionality to idempotent'ly reach a target state on a system by invoking 'module' actions. Imperative, in a way. In Ansible terms: task implementation I believe.
- State SLS: combination/invocation of state procedures making up a final state. Very declarative. In Ansible terms: role I believe.

- Runner procedure: invoked in context of the master, can invoke module procedures on master or minions, and other runners (again invoked in master context). Used to orchestrate things, mainly. In Ansible terms: playbook I believe.
- `state.orchestrate` runner: special case of a `runner` procedure which uses an SLS as 'input' and interprets it as a DSL instead of requiring one to write the Python code of the same effect.

TeddyAndrieux replied:
I agree but for me the goal was that the master is the only "salt" who talk to the apiserver, No ?

NicolasT replied:
- Currently this code goes through the API-server because it uses an `exec` in an `etcd` Pod to perform a certain action. That's not the right approach, cfr. #878.
- Even when #878 gets implemented, the call would still happen from the master because that's the only one who's supposed to have the right credentials to invoke the command.
- This is a step which should be executed 'once', so it's part of 'orchestration'.

However.

Whilst we may invoke this procedure on the master, and as part of a larger playbook/orchestration, that doesn't mean the functionality itself (i.e. the procedure) couldn't be useful in other contexts. So, it's a module procedure, which may be invoked from an orchestration/runner, or in a state module if ever, or ...

In general:

- An 'action', imperative procedure to affect 'real world': module procedure.
- An 'orchestration', steps to be followed with centralized command (~ running on the master): runner procedure (or `state.orchestrate` SLS, which is kinda the same).

The question of 'where do *we* currently execute this code' is not part of the equation.

TeddyAndrieux replied:
Ok make sense

gdemonet reviewed (commented):
So if I understood correctly, we'd like to invoke the orchestrate like so:

```
salt-run state.orchestrate metalk8s.orchestrate.deploy_etcd_on_new_node saltenv=metalk8s-2.0 --pillar='{"node_name": "node1", "node_ip": "10.0.0.2"}'
```

Could we use the Node object from the API server to retrieve the node IP instead? Maybe not for this PR though.

slaperche-scality commented:
> So if I understood correctly, we'd like to invoke the orchestrate like so:
> 
> ```
> salt-run state.orchestrate metalk8s.orchestrate.deploy_etcd_on_new_node saltenv=metalk8s-2.0 --pillar='{"node_name": "node1", "node_ip": "10.0.0.2"}'
> ```

Yup that's it!


> Could we use the Node object from the API server to retrieve the node IP instead? Maybe not for this PR though.

Not sure to understand, what would be the required changes?

slaperche-scality commented:
PR updated with the last fixes from the live debug done with @nootal, @ramanduh and @sayf-eddine-scality.

To be tested, this PR should be rebased on top of @nootal  and @ramanduh PRs.

slaperche-scality commented on salt/metalk8s/orchestrate/deploy_etcd_on_new_node.sls:
```diff
@@ -0,0 +1,81 @@
+      - salt: Generate etcd certificates on a new node
+      - module: Refresh the list of etcd nodes
+
+Deploy updated manifests for existing etcd nodes:
```
Hum, really not sure about this one.

The more i think about it, the more it seems wrong.
I'll redo tests without it, looks like a wild guess made during our troubleshooting session.

NicolasT replied:
I don't think there's a need for this indeed?

sayf-eddine-scality replied:
agree there is no need to touch existing members

slaperche-scality replied:
Indeed, I confirmed that it was useless.

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_etcd_on_new_node.sls:
```diff
@@ -0,0 +1,66 @@
+Refresh the list of etcd nodes:
+  module.run:
+    - mine.update:
+      - mine_functions: metalk8s.get_etcd_endpoint
```
This will run the `mine.update` on the master which is basically not what we want.

`mine.update` it's a minion that push information to the master (NOT master who get information from all minions) so you should call mine.update on all salt-minions.

Something like that
```
Refresh the list of etcd nodes:
  salt.function:
    - name: mine.update
    - tgt: '*'
```

`mine_functions` is to update only specific informations maybe we don't care here.
Me personnaly I don't care update all mine or only the one you need, as you want :)

NicolasT replied:
I think we're trying to fix an issue we shouldn't have in the first place because the mine is not the right place to figure out the layout of the etcd cluster (cfr. #877).

slaperche-scality replied:
Ha that may explain why it wasn't working, thanks for the explanation @TeddyAndrieux.

gdemonet commented:
> > Could we use the Node object from the API server to retrieve the node IP instead? Maybe not for this PR though.
>
> Not sure to understand, what would be the required changes?

My point was that we did not need to pass `"node_ip": "..."` as pillar overrides when running this orchestrate, the `node_name`, which is also the minion ID, should be self-sufficient.

I don't think it's used in your code anyway ([in salt/metalk8s/orchestrate/deploy_etcd_on_new_node.sls](https://github.com/scality/metalk8s/pull/869/files#diff-1392062438ad991f95f2192e0e1af4f9)):
```
{%- if pillar['node_name'] in control_plane_ips.keys() and control_plane_ips[pillar['node_name']] %}
{%- set node_ip = control_plane_ips[pillar['node_name']][0] %}
{%- else %}
{%- set node_ip = 'localhost' %}
{%- endif %}
```

slaperche-scality commented:
> I don't think it's used in your code anyway

Yup, this part was reworked and I no longer need the IP to be passed in the pillar.

TeddyAndrieux commented on salt/_modules/metalk8s.py:
```diff
@@ -0,0 +1,87 @@
+def __virtual__():
+    '''Check dependencies.'''
+    if HAS_DEPS:
+        return HAS_DEPS
```
More common
```suggestion
        return __virtualname__
```

slaperche-scality replied:
Done

slaperche-scality commented on salt/metalk8s/orchestrate/deploy_etcd_on_new_node.sls:
```diff
@@ -0,0 +1,66 @@
+
+Refresh the list of etcd nodes:
+  salt.function:
+    - name: mine.update
```
Didn't find the right syntax to only update the `etcd` endpoints, so I refresh the whole mine…

TeddyAndrieux replied:
Should be something like
```
Refresh the list of etcd nodes:
  salt.function:
    - name: mine.update
    - tgt: '*'
    - kwarg:
        mine_function: etcd_endpoints
```
But honnestly for me it's ok to refresh all the mine as we will likely not use mine in the futur for this

slaperche-scality replied:
I think I tried that (maybe I used `kwargs` instead of `kwarg` though) and I got Salt complaining about me doing dirty things that won't work with Salt Fluorine or sth like that.

Agreed, since we won't rely on the mine for that in the future, let's keep it as-is.

slaperche-scality commented:
@bert-e approve

bert-e commented:
# History mismatch



Merge commit #160fe53fe8c4c244200a1ff0fe745dbcc1c698ce on the integration branch
`w/2.1/feature/738-etcd-cluster-expansion` is merging a branch which is neither the current
branch `feature/738-etcd-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/738-etcd-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/738-etcd-cluster-expansion` into
`development/2.0`
* [w/2.1/feature/738-etcd-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/738-etcd-cluster-expansion) will be merged into `development/2.1` 
* [w/2.2/feature/738-etcd-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/738-etcd-cluster-expansion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

NicolasT commented on salt/_modules/metalk8s.py:
```diff
@@ -39,3 +54,58 @@ def get_workload_plane_ips():
+            raise CommandExecutionError('etcdctl: {}'.format(err))
+    except ApiException as exn:
+        log.exception('failed to run etcdctl')
+        raise exn
```
Just `raise`, not `raise exn`, right, to keep correct backtrace?

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/62112dfee0fb67e2fad5407df96e958388283bd5) did not succeed in branch w/2.2/feature/738-etcd-cluster-expansion.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
